Introduce `fallthrough` as a wrapper around the fallthrough attribute

MisterDA opened the pull request:
Recent compilers have introduced the `-Wimplicit-fallthrough` warning
against the C switch fallthrough. OCaml already uses the `/* fallthrough */`
 C comment that GCC 7 and above recognize for some level
given to the warning, but the fallthrough statement is also available
since C++17 and either as a language statement (expected in C2X) or as
a compiler/language attribute.

This commit makes use of the statement/attribute when available, or
falls back to the fallthrough comment, and uses the new
`CAMLfallthrough` macro wherever the comment was used.

MisterDA committed:
Introduce fallthrough as a wrapper around fallthrough attribute

Recent compilers have introduced the -Wimplicit-fallthrough warning
against the C switch fallthrough. OCaml already uses the
/* fallthrough */ C comment that GCC 7 and above recognize for some
level given to the warning, but the fallthrough statement is also
available since C++17 and either as a language statement (expected in
C2X) or as a compiler/language attribute.

This commit makes use of the statement/attribute when available, with
a fallback to an empty statement, and uses the new fallthrough macro
wherever the comment was used.

It also reorders some switches to avoid "control reaches end of
non-void function" warnings when CAMLassert is disabled.

xavierleroy commented on runtime/caml/misc.h:
```diff
@@ -181,6 +192,16 @@ CAMLdeprecated_typedef(addr, char *);
+#elif caml_has_attribute (fallthrough)
+  #define CAMLfallthrough __attribute__ ((fallthrough))
+#else
+  #define CAMLfallthrough /* fallthrough */
```
I'm pretty sure this macro will expand to nothing, not to the `/* fallthrough */` comment.  Did you test this case?

MisterDA replied:
You're right, I didn't notice that macro expands to nothing, the preprocessor removes comments in macros. I can't think of a way to provide a fallback to the `/* fallthrough */` comment.

gasche replied:
Expanding to nothing is fine, as the macro itself plays the readability role of the comment, and I don't think anyone is reading our post-processed sources? (The comment itself can/should probably be removed from the macro definition.)

nojb replied:
Note that gcc can recognize the "fallthrough" *comment*, see the doc for `-Wimplicit-fallthrough=3` at https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html

MisterDA replied:
Thank you for the comments, changed the fallback definition to using a noop statement `#define CAMLfallthrough ((void) 0)`.

MisterDA commented:
Found more cases of fallthrough to replace.
Shoud I probe the compiler for support of `-Wimplicit-fallthrough` in the configure script to enable it by default?
Note that in gcc, it's also part of `-Wextra`.

xavierleroy commented:
I'm not sure what you're trying to achieve.  For quite a while, gcc has been able to check `switch` statements based on the presence of `/* fallthrough */` comments, plus the appropriate -Wimplicit-fallthrough flag.  Why do you want to replace nice-looking `/* fallthrough */` comments  by scary-looking `CAMLfallthrough` macro and its scary-looking, `#ifdef`-fest definition?  What is gained?

MisterDA commented:
I agree that comments in general look nicer than the attributes or my
`CAMLfallthrough` proposal. 

I think however that using the attributes/statement syntax has
advantages over the comments. Some of the reasons below are more
subjective than others, they may be of less relevance to you.

I'm unconfortable putting a semantic meaning into a comment. I think
they should be reserved to the programmer. Having the compiler peek
into them looks at worst flaky to me, at best the poor man's static
analysis. If we were to run the preprocessor separately from the
compiler, we'd have to tell it to preserve comments. This is also very
much dependent on the compiler. Clang doesn't read comments but
supports the attribute. The fallthrough statement has been added to
C++17 and it has been included in the draft of the next C
revision. Soon-ish we'll be able to make use of an actual language
feature rather than a compiler extension (or a compiler reading the
comment section).

I also don't find it obvious which level to use for the warning; in my
tests GCC 11.2 (mingw-w64) there always was at least one warning (in
globroots.c or interp.c) regardless of the level, even if there was a
variation on the `fallthrough` comment. Instead of fixing this comment
so that it can be recognized by a specific compiler, we could change
the code so that it can have the best relationship with all
compilers.

In interp.c I prefer `Fallthrough;` for the parallel with `Next;`
rather than variations on `/* Fallthrough */`.

Now, to get a better looking macro we could use simply `fallthrough;`
as done in the Linux kernel, and we could put it in a private header
not installed, to drop the __caml__ prefix namespace. I think that
would also simplify the macro dance.

I haven't triggered any warning report for a forgotten fallthrough
comment, indicating that the code is well annotated and not in dire
need of this patch (but perhaps a slight adjustment to some comments),
so if you're not convinced I won't push for it further.

MisterDA commented:
Rebased the PR to fix Changes conflicts.
Let me know what I can do to improve it, if needed.

MisterDA committed:
Use C23 __has_c_attribute and __has_attribute extension

`__has_c_attribute` is C23 and checks for `[[name]]`,
`__has_attribute` is an extension of the compiler and checks for
`__attribute__((name))`.

dra27 commented:
This PR would have allowed us to avoid the slip in #11002 fixed in #11066. https://github.com/dra27/ocaml/commit/4882f3e09b506ddea41b3b2e482731faeae357ab adds the necessary support to `configure` to add `-Wimplicit-fallthrough` to the C warnings enabled during the build (I wish to atone for failing to spot the issue when I reviewed #11002, even if I did spot another bug...).

There is a case for the `CAMLfallthrough` macro vs having the comments: I believe clang supports `-Wimplicit-fallthrough` but doesn't recognise gcc's `/* fall through */` markers.

MisterDA commented:
This PR (or at least the warning) would have caught this new bug before it was introduced, maybe we could revisit it? Move the detection of the attribute to configure as not to clutter the headers?

dra27 commented:
`runtime/interp.c` creates a subtle problem if we just use the existing comments approach. I'm guessing that whatever GCC does with the comments must occur _before_ pre-processing, which means that it doesn't realise that  `Instruct(ACC)` is actually a `case` at the time it looks at the `/* fallthrough */` comment. The warning therefore triggers on `interp.c` when compiled for the debug runtime (which is not threaded) regardless of the comments.

However, it turns out this can be fixed by including `-CC` (which keeps comments during pre-processing). I have a minimal https://github.com/dra27/ocaml/commit/f425a4e02bc2de32bf190a23cb1dee79b3d9bdb4 which adjusts just the two "fall through" comments which GCC doesn't recognise and turns the warning on for GCC only.

I think we should definitely merge one of these solutions, although my preference is for @MisterDA's, as it enables the warning for clang as well.

MisterDA committed:
Turn on -Wimplicit-fallthrough when available

MisterDA commented:
I've cleaned, simplified, and rebased this PR. Could this be reconsidered?

MisterDA commented:
> if you're not convinced I won't push for it further.

Evidently, I lied.

> There is a case for the `CAMLfallthrough` macro vs having the comments: I believe clang supports `-Wimplicit-fallthrough` but doesn't recognise gcc's `/* fall through */` markers.

That's the strongest point, I think.

I've rebased once more, and chosen to guard definition of `fallthrough` under `CAML_INTERNALS`, so that there's no need to namespace it.

gasche commented on runtime/interp.c:
```diff
@@ -64,6 +64,8 @@ sp is a local copy of the global variable Caml_state->extern_sp. */
 #    define Next goto *(void *)(jumptbl_base + *pc++)
 #  endif
+#  undef fallthrough
+#  define fallthrough ((void) 0)
```
Nitpick: I find it arcane to change the semantics of `fallthrough` in just this file, I would rather introduce a `Continue` macro (or `Fallthrough` or something else than `fallthrough`) that is defined as `(void) 0` in the threaded case and `fallthrough` otherwise.

MisterDA replied:
That's a good point. I'll reuse `Fallthrough` in this file.

gasche reviewed (approved):
I don't have a clear opinion or expertise on these topics, but I can see that @MisterDA would be happy to get some feedback so I had a look. This looks okay: I trust @dra27's positive gut feeling, and I agree with the sentiment that proper syntax is better than a comment. I don't see a cost to this PR besides the churn and potential for regression, and arguably some benefits, so I propose to go ahead and merge it eventually.
